zcash_primitives/src/zip32.rs

```diff
@@ -43,41 +48,148 @@ impl ConditionallySelectable for AccountId {
+/// not.
+pub trait Hardening: private::Sealed {}
+
+mod private {
+    pub trait Sealed: Sized {
+        fn master() -> Self;
+        fn hardened(index: u32) -> Self;
+        fn is_hardened(&self) -> bool;
+        fn index(&self) -> u32;
+    }
+}
+
+/// Marker indicating that a [`ChildIndex`] is always hardened.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Hard(u32);
+
+impl private::Sealed for Hard {
+    fn master() -> Self {
+        // Hardened-only derivation still uses the non-hardened child index of 0 for the
+        // master key, for historic reasons. This only appears in the extended spending
+        // key encoding format, and is not used anywhere.
+        Self(0)
+    }
+
+    fn hardened(index: u32) -> Self {
+        Self(index)
+    }
+
+    fn is_hardened(&self) -> bool {
+        true
+    }
+
+    fn index(&self) -> u32 {
+        self.0
+    }
+}
+
+impl Hardening for Hard {}
+
 #[derive(Clone, Copy, Debug, PartialEq, Eq)]
-pub enum ChildIndex {
+enum FlexibleKind {
     NonHardened(u32),
-    Hardened(u32), // Hardened(n) == n + (1 << 31) == n' in path notation
+    Hardened(u32),
 }
 
-impl ChildIndex {
-    pub fn from_index(i: u32) -> Self {
-        match i {
-            n if n >= (1 << 31) => ChildIndex::Hardened(n - (1 << 31)),
-            n => ChildIndex::NonHardened(n),
+/// Marker indicating that a [`ChildIndex`] can be either hardened or non-hardened.
+///
+/// This is a legacy marker for transparent and Sapling usage. Non-hardened derivation is
+/// unsupported by Orchard and subsequent protocols.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Flexible(FlexibleKind);
+
+impl private::Sealed for Flexible {
+    fn master() -> Self {
+        Self(FlexibleKind::NonHardened(0))
+    }
+
+    fn hardened(index: u32) -> Self {
+        Self(FlexibleKind::Hardened(index))
+    }
+
+    fn is_hardened(&self) -> bool {
+        matches!(self.0, FlexibleKind::Hardened(_))
+    }
+
+    fn index(&self) -> u32 {
+        match self.0 {
+            FlexibleKind::NonHardened(i) => i,
+            FlexibleKind::Hardened(i) => i,
         }
     }
+}
 
-    fn master() -> Self {
-        ChildIndex::from_index(0)
+impl Hardening for Flexible {}
+
+/// A child index for a derived key.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct ChildIndex<H: Hardening>(H);
+
+impl<H: Hardening> ChildIndex<H> {
+    /// Constructs the sentinel `ChildIndex` to use for master keys (that are not derived
+    /// child keys).
+    pub fn master() -> Self {
+        Self(H::master())
     }
 
-    fn value(&self) -> u32 {
-        match *self {
-            ChildIndex::Hardened(i) => i + (1 << 31),
-            ChildIndex::NonHardened(i) => i,
+    /// Constructs a hardened `ChildIndex` from the given value.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `value >= (1 << 31)`.
+    pub fn hardened(value: u32) -> Self {
```

Comment:
Why panic instead of return `Option<Self>`?

Reply:
This is for simplicity, as its use case is basically always with constants. The API now has `ChildIndex::from_index` again, which is now fallible.

Reply:
Can this now be `pub const`?